### PR TITLE
#3718 delta sync pull

### PR DESCRIPTION
**Pull request by bbrks**

### CODE REVIEW

File: rest/blip_api_test.go [hunk 1]
```diff
@@ -1475,3 +1480,34 @@ func TestMissingNoRev(t *testing.T) {
...
+	assert.True(t, ok)
+
+	// doc1 rev 2-959f0e9ad32d84ff652fb91d8d0caa7e
+	resp = bt.restTester.SendAdminRequest(http.MethodPut, "/db/doc1?rev=1-0335a345b6ffed05707ccc4cbc1b67f4", `{"greetings": [{"hello": "world!"}, {"hi": "alice"}, {"howdy": "bob"}]}`)
+	assert.Equal(t, http.StatusCreated, resp.Code)
+
+	ok, err = client.WaitForRev("doc1", "2-959f0e9ad32d84ff652fb91d8d0caa7e")
+	assert.NoError(t, err)
+	assert.True(t, ok)
+}
```
Comment (bbrks) [inline rest/blip_api_test.go:1540]: Need to assert that the stored revision is the full doc, not a delta.

File: rest/blip_sync.go [hunk 2]
```diff
@@ -464,7 +474,15 @@ func (bh *blipHandler) handleChangesResponse(sender *blip.Sender, response *blip
 					return
 				}
 			}
-			bh.sendRevOrNorev(sender, seq, docID, revID, knownRevs, maxHistory)
+
+			fmt.Printf("  DELTAS: bh.useDeltas was: %v\n", bh.useDeltas)
```
Comment (bbrks) [inline rest/blip_sync.go]: remove debug comments

File: rest/blip_client_test.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+			// changesReqs == [[sequence, docID, revID, {deleted}, {size (bytes)}], ...]
+			for i, changesReq := range changesReqs {
+				docID := changesReq[1].(string)
+				btc.docsLock.RLock()
+				if revs, haveDoc := btc.docs[docID]; haveDoc {
+					revList := make([]string, 0, len(revs))
+					for revID := range revs {
+						revList = append(revList, revID)
+					}
+					wantedRevs[i] = revList // send back all revs we have for SG to determine the ancestor
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: If we already have the rev from the changes message in revList, indicate in response that we don't need this change.

Comment (adamcfraser) [inline rest/blip_client_test.go]: Stick a TODO: in here to limit the number of wantedRevs (knownRevs) to 20, as I think that's the CBL behaviour.

File: rest/blip_client_test.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+			err = json.Unmarshal(body, &changesReqs)
+			if err != nil {
+				panic(err)
+			}
+
+			wantedRevs = make([]interface{}, len(changesReqs))
+			// changesReqs == [[sequence, docID, revID, {deleted}, {size (bytes)}], ...]
+			for i, changesReq := range changesReqs {
+				docID := changesReq[1].(string)
+				btc.docsLock.RLock()
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: TODO about a future enhancement to push locking down to a set of accessors on btc.

Comment (adamcfraser) [inline rest/blip_client_test.go]: Move RLock outside the loop to simplify (for future locking refactoring)

File: rest/blip_client_test.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+					}
+					wantedRevs[i] = revList // send back all revs we have for SG to determine the ancestor
+				} else {
+					wantedRevs[i] = []interface{}{} // sending empty array means we've not seen the doc before, but still want it
+				}
+				btc.docsLock.RUnlock()
+			}
+		}
+
+		if !request.NoReply() {
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: Can we check request.NoReply at the top of the handler and exit early?

File: rest/blip_client_test.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		docID := request.Properties["id"]
+		revID := request.Properties["rev"]
+
+		body, err := request.Body()
+		if err != nil {
+			panic(err)
+		}
+
+		// If deltas are enabled, and we see a deltaSrc property, we'll need to patch it before storing
+		if btc.Deltas && bytes.Contains(body, []byte(`"deltaSrc":`)) {
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: Rework to send deltaSrc as a request property.

File: rest/blip_client_test.go [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+			delete(old, "deltaSrc") // remove deltaSrc after patching
+
+			body, err = json.Marshal(old)
+			if err != nil {
+				panic(err)
+			}
+		}
+
+		btc.docsLock.Lock()
+		btc.docs[docID] = map[string][]byte{revID: body}
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: If doc already exists, append new rev instead of rewriting map.

File: rest/blip_client_test.go [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	bt.blipContext.HandlerForProfile[messageGetAttachment] = func(request *blip.Message) {
+		btc.Requests <- request
+		panic("getAttachment not implemented")
+	}
+
+	bt.blipContext.DefaultHandler = func(request *blip.Message) {
+		btc.Requests <- request
+		log.Printf("defaultHandler got request: %v", request)
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: As discussed, panic here as it's unexpected behaviour.

File: rest/blip_client_test.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	if !btc.bt.sender.Send(getCheckpointRequest) {
+		log.Fatalf("unable to send getCheckpointRequest")
+	}
+
+	subChangesRequest := blip.NewRequest()
+	subChangesRequest.SetProfile(messageSubChanges)
+	subChangesRequest.Properties[subChangesContinuous] = "true"
+	subChangesRequest.Properties[subChangesSince] = "0"
+	subChangesRequest.Properties[subChangesFilter] = "sync_gateway/bychannel"
+	subChangesRequest.Properties[subChangesChannels] = "*"
```
Comment (adamcfraser) [inline rest/blip_client_test.go]: Can omit channel filter for the simple case (will replicate all channels the user has access to).  Could stub out a StartChannelFiltered() API?

**Review by adamcfraser — DISMISSED** [review]

File: rest/blip_api_test.go [hunk 10]
```diff
@@ -1475,3 +1480,42 @@ func TestMissingNoRev(t *testing.T) {
...
+
+	data, ok := client.WaitForRev("doc1", "1-0335a345b6ffed05707ccc4cbc1b67f4")
+	assert.True(t, ok)
+	assert.Equal(t, `{"greetings":[{"hello":"world!"},{"hi":"alice"}]}`, string(data))
+
+	// create doc1 rev 2-959f0e9ad32d84ff652fb91d8d0caa7e
+	resp = bt.restTester.SendAdminRequest(http.MethodPut, "/db/doc1?rev=1-0335a345b6ffed05707ccc4cbc1b67f4", `{"greetings": [{"hello": "world!"}, {"hi": "alice"}, {"howdy": "bob"}]}`)
+	assert.Equal(t, http.StatusCreated, resp.Code)
+
+	// TODO: assert that this change was sent as a delta
```
Comment (bbrks) [inline rest/blip_api_test.go]: TODO: assert that this change was sent as a delta in EE, and a full doc in CE

File: manifest/default.xml [hunk 11]
```diff
@@ -117,6 +117,10 @@
 
   <project name="testify" path="godeps/src/github.com/stretchr/testify" remote="couchbasedeps" revision="04af85275a5c7ac09d16bb3b9b2e751ed45154e5"/>
 
+  <!-- Enterprise edition dependencies -->
+  <project groups="notdefault,cb_sg_enterprise" name="go-fleecedelta" path="godeps/src/github.com/couchbaselabs/go-fleecedelta" remote="couchbaselabs_private" revision="1277ab2af4e7b861a66aadca51a7f49679f9522f"/>
+  <project groups="notdefault,cb_sg_enterprise" name="go-diff" path="godeps/src/github.com/sergi/go-diff" remote="couchbasedeps" revision="da645544ed44df016359bd4c0e3dc60ee3a0da43"/>
+
```
Comment (bbrks) [inline manifest/default.xml:123]: Currently these only get pulled when the project is set to `sg-accel`

We might need to decide if the bootstrap script should take an edition parameter instead?

File: rest/blip_sync.go [hunk 12]
```diff
@@ -553,6 +568,73 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
...
+		bh.sendRevOrNorev(sender, seq, docID, revID, knownRevs, maxHistory)
+		return
+	}
+
+	revsStart := strconv.Itoa(newBodyRevisions[db.RevisionsStart].(int) - 1) // previous rev start
+	for _, v := range newBodyRevisions[db.RevisionsIds].([]string) {
+		rev := revsStart + "-" + v
+		bh.Logf(base.LevelTrace, base.KeySync, "DELTA:   comparing ancestor rev: %v to knownRevs: %v", rev, knownRevs)
+		if knownRevs[rev] {
+			bh.Logf(base.LevelInfo, base.KeySync, "DELTA: matched ancestor rev to known rev: %v", rev)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: Too verbose for Info logging.

File: rest/blip_sync.go [hunk 13]
```diff
@@ -553,6 +568,73 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
...
+			deltaSrcRevID = rev
+		}
+	}
+
+	bh.Logf(base.LevelDebug, base.KeySync, "DELTA: getting oldBody docID: %v - rev: %v", docID, deltaSrcRevID)
+	oldBody, err := bh.db.GetRev(docID, deltaSrcRevID, true, nil)
+	if err != nil {
+		bh.Logf(base.LevelInfo, base.KeySync, "DELTA: couldn't get oldBody - falling back to full body replication... err: %v", err)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: Too verbose for Info logging.

File: rest/blip_sync.go [hunk 14]
```diff
@@ -553,6 +568,73 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
...
+		// fall back to standard full-body replication if we can't do a delta
+		bh.sendRevOrNorev(sender, seq, docID, revID, knownRevs, maxHistory)
+		return
+	}
+
+	// generate delta
+	bh.Logf(base.LevelTrace, base.KeySync, "DELTA: generating delta... oldBody, newBody")
+	delta, err := base.Diff(oldBody, newBody)
+	if err != nil {
+		bh.Logf(base.LevelWarn, base.KeySync, "DELTA: couldn't generate delta - falling back to full body replicaiton... err: %v", err)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: Typo in replication

**Review by adamcfraser — DISMISSED** [review]

**Review by adamcfraser — APPROVED** [review]
Approved - waiting to merge until we have integration/PR build support for new private repo.

**Pull request merged**